### ISSUE TITLE
GetTwin Method Misapplies 5 nesting depth limit for module and device twin

### ISSUE THREAD

**Issue by sam-findler** [Azure/azure-iot-sdk-java#338]
- **OS and version used:** Ubuntu 14.04

- **Java runtime used:** jdk 8u171

- **SDK version used:** 1.14.0

# Description of the issue:
Acccording to https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-device-twins#tags-and-properties-format , the nesting level inside of the reported_properties, desired_properties and tags is supposed to be 5, starting from the inside of the map.  However, the code in the client (https://github.com/Azure/azure-iot-sdk-java/blob/master/deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java) indicates that the nesting level is being counted from the top level.  We have not seen this issue in the c/python client, only in the java client.

# Code sample exhibiting the issue:

` 
//creating the property
private static Property createAlertProperty(int dataIndex) {
    Map<String, Object> alertProperties = new HashMap<>();
    Map<String, Object> sourceProperties = new HashMap<>();
    sourceProperties.put(TwinFields.OK.value(), 1);
    Map<String, Object> data = new HashMap<>();
    data.put("data_index", dataIndex);
    sourceProperties.put(TwinFields.DATA.value(), data);
    alertProperties.put("some_source_1", sourceProperties);
    alertProperties.put(TwinFields.PRIORITY.value(), 1);

    return new Property("some_alert_1", alertProperties);
  }
/*
This produces the reported property:
"some_alert_1":{  
            "some_source_1":{  
               "data":{  
                  "data_index":0
               },
               "ok":1
            },
            "priority":1
 }
Which is only a nesting level of 3 deep
*/
`
`
//sending the reported property
Property property = createAlertProperty(i);
System.out.println("Sending Property" + property.getKey());
moduleClient.sendReportedProperties(new HashSet<>(Arrays.asList(property)));
`
`
//trying to get the remote device twin state from the service client

DeviceTwin service = DeviceTwin
        .createFromConnectionString(iotHubTestController.getIotHubConnectionString());
DeviceTwinDevice device = new DeviceTwinDevice(moduleController.getModule().getDeviceId(),
        moduleController.getModule().getId());
//this is when the exception occurs
service.getTwin(device);
`

# Console log of the issue:
`
java.lang.IllegalArgumentException: Map exceed maximum of 5 levels

	at com.microsoft.azure.sdk.iot.deps.serializer.ParserUtility.validateMapInternal(ParserUtility.java:237)
	at com.microsoft.azure.sdk.iot.deps.serializer.ParserUtility.validateMapInternal(ParserUtility.java:233)
	at com.microsoft.azure.sdk.iot.deps.serializer.ParserUtility.validateMapInternal(ParserUtility.java:233)
	at com.microsoft.azure.sdk.iot.deps.serializer.ParserUtility.validateMapInternal(ParserUtility.java:233)
	at com.microsoft.azure.sdk.iot.deps.serializer.ParserUtility.validateMapInternal(ParserUtility.java:233)
	at com.microsoft.azure.sdk.iot.deps.serializer.ParserUtility.validateMap(ParserUtility.java:205)
	at com.microsoft.azure.sdk.iot.deps.twin.TwinCollection.put(TwinCollection.java:231)
	at com.microsoft.azure.sdk.iot.deps.twin.TwinCollection.put(TwinCollection.java:96)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:188)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:145)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
	at com.google.gson.Gson.fromJson(Gson.java:887)
	at com.google.gson.Gson.fromJson(Gson.java:852)
	at com.google.gson.Gson.fromJson(Gson.java:801)
	at com.google.gson.Gson.fromJson(Gson.java:773)
	at com.microsoft.azure.sdk.iot.deps.twin.TwinState.createFromTwinJson(TwinState.java:297)
	at com.microsoft.azure.sdk.iot.service.devicetwin.DeviceTwin.getTwinOperation(DeviceTwin.java:113)
	at com.microsoft.azure.sdk.iot.service.devicetwin.DeviceTwin.getTwin(DeviceTwin.java:88)
`

# Need Support?
* Have a feature request for SDKs? Please post it on [User Voice](https://feedback.azure.com/forums/321918-azure-iot) to help us prioritize.
* Have a technical question? Ask on [Stack Overflow](https://stackoverflow.com/questions/tagged/azure-iot-hub) with tag “azure-iot-hub”
* Need Support? Every customer with an active Azure subscription has access to support with guaranteed response time.  Consider submitting a ticket and get assistance from Microsoft support team
* Found a bug? Please help us fix it by thoroughly documenting it and filing an issue on GitHub (C, Java, .NET, Node.js, Python).

**Comment by timtay-microsoft** [comment]
Thanks for filing this issue. We'll try to take a look into it soon, but we'd also be happy to review a PR for you if you'd like.

**Comment by timtay-microsoft** [comment]
You are correct that for Desired/reported properties, the max depth needs to be raised. We'll try to look into this soon. The current depth limit only works for Tags right now.

**Comment by sam-findler** [comment]
The current max depth also doesn't work for Tags, though it is only 1 off for tags instead of 2 off for desired/reported properties.  The current model assumes the tag name at the top level instead of "tags:" { "tag_name": ... }.  I could rewrite the PR to increase the depth by 1 instead of 2 in lieu of a full model rewrite capable of differentiating between types of data in the twin.

**Comment by timtay-microsoft** [comment]
It may be simpler for us to just remove this check altogether. The service will perform the appropriate check, and return a descriptive error message if the depth is too large. I'll need to discuss this approach with the rest of the team.

I'd rather not just increase the depth by 1.

**Comment by sam-findler** [comment]
That seems fair to me, it is redundant to do the depth check client side as it stands now.

**Comment by timtay-microsoft** [comment]
Sorry for the delay on this issue, we have a fix for this issue in the works. I'll update this thread when it gets checked into master.

**Comment by az-iot-builder-01** [comment]
@sam-findler, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
